### ISSUE TITLE
Extreme slowdown at specific execution timesteps

### ISSUE THREAD

**Issue by reint-fischer** [OceanParcels/parcels#746]
Hi all,

I am releasing particles in an open 3D domain with velocities only in the [x,z]-plane.

I am running into a sudden extreme increase in execution time at specific moments for specific combinations of dt and outputdt. At different starttimes, fieldsets, and both in forwards in time and backwards in time execution parcels takes extremely long at the same specific timestep. 

Decreasing dt seems to make this moment come earlier, while decreasing outputdt seems to do the opposite. However, some values for dt work significantly better, i.e. no difference between 0.005 and 0.01 but large difference between 0.005 and 0.004.

A number of combinations I have found:
DT [s]                      OutputDT [s]                     Timestep [s] where parcels starts taking very long
0.002                      0.1                                     0.6
0.002                      0.08                                   0.64
0.002                      0.01                                   5.42
0.004                      0.1                                     0.9999
0.004                      0.08                                   1.28
0.005                      0.1                                     44.8
0.005                      0.08                                   44.8
0.01                        0.1                                     44.8
0.01                        0.08                                   44.8

Since the bug is irrespective of starttime or fieldset, I do not think anything is happening to the particles. My memory looks fine. I get no errors. The problem is exactly reproducible.

[ExtremeSlowdown.txt](https://github.com/OceanParcels/parcels/files/4197095/ExtremeSlowdown.txt)

Any ideas what this might be?

**Comment by reint-fischer** [comment]
Hi, as a quick fix for the bug, we've tried increasing the `__tol` variable. This seems to increase the time it takes before the extreme slowdown occurs and perhaps even speed up the continuation. It doesn't get rid of the problem altogether however. 

Now, I have found that for `__tol` == 1e-4, outputdt = 0.1 and dt = 0.001 I can get Parcels to execute for 44 seconds, with a slowdown at 21.3 seconds. After the slowdown, it seems something changes in the outputdt, as you can see in these figures:
![bug1zoom](https://user-images.githubusercontent.com/33000433/75148762-b2d2d700-5700-11ea-8d4b-f56a212e4d5f.png)
![bug1](https://user-images.githubusercontent.com/33000433/75148763-b36b6d80-5700-11ea-9277-aa9deb2f9cce.png)

Up until 21.3 seconds, each timestep takes 0.1 seconds, so the line is nicely 1:1. After the slowdown, the dt becomes variable. You can see this in this figure: 
![bug1dt](https://user-images.githubusercontent.com/33000433/75148988-3e4c6800-5701-11ea-848e-d5f6c8b09bdc.png)

The outputdt starts at 100 ms = 0.1 s, but after 21.3 seconds it hops around from 100 to 400 to 100 to 300.

Christian mentioned this bug might be related to the pull-request  #751 .

**Comment by reint-fischer** [comment]
I have been looking at it some more and I have two things that I'd like to check:

1. For now, we have only changed something in the JIT-version of the execution loop, should we also change something in the python-version?

2. The tolerance variable is compared to dt_pos in the python-version and __dt in the JIT-version. These are calculated by taking the minimum of p.dt and the absolute difference between endtime and p.time. Is it possible that the bug is caused by p.time overshooting endtime by more than the tolerance, creating an infinite while-loop? And if so, could we not change the definition of dt_pos so that the while-loop also stops when p.time is on the 'wrong' side of endtime?

Feel free to ask me for clarification of my thoughts

**Comment by CKehl** [comment]
Actually, let's look at line 942 - 944:
```
        dt_pos = c.Assign("__dt", "fmin(fabs(particles[p].dt), fabs(endtime - particles[p].time))")
        pdt_eq_dt_pos = c.Assign("__pdt_prekernels", "__dt * sign_dt")
        partdt = c.Assign("particles[p].dt", "__pdt_prekernels")
```

Problem: `p.dt` is not always `pset.dt`, because it is not reset for erroneous loop executions, and this offset error accumulates in `p.time`.

**Comment by CKehl** [comment]
What I'm thinking here (and first test in scipy and then here): semantic problem of the `ErrorCode.Success` - the code expresses (a) all particles that still need to be computed per time step; (b) all particles that are (already) successfully calculated; (c) all particles that need to be re-calculated. If we would actually split that up, we could (in the particle for-loop and as pre-condition to the while-loop for kernel execution) actually say to only (re-)calculate particles that need to be calculated - and ignore additional kernel evaluations. That would, semantically to me, also be the expected norm.

It does not prevent re-calculation in the future ... if you bind particles in groups and, for some user-defined kernel behaviour and reason, particles need to be re-evaluated, then the user kernel could manually reset this 'ErrorCode' (which, at that point, is more of a 'StatusCode') at will.

Still, the norm (and expectation, I guess) is: If a particle is, for one dt time iteration, evaluated correctly (without error), do not re-evaluate it.

**Comment by erikvansebille** [comment]
Yes, I think that implementing this semantic change is a very good idea. Feel free to implement a proof of concept?

**Comment by CKehl** [comment]
fixed by #762

**Comment by reint-fischer** [comment]
Hey, @cjongedijk and I have done some more testing, since we've figured out that this problem hasn't really been solved after all. First we found out that p.dt becomes really small at the extreme slowdown (1E-4 -> 1E-8). Then, we figured out that the slowdown is caused by the fact that this dt stays 1E-8 throughout the rest of the simulation and therefore 1E4 more computations have to be carried out.

Next we tested how the p.dt resets after finishing the kernel.execute loop. I.e. we ran a simulation with dt=0.03 and outputdt=0.1. For the first three timesteps p.dt=0.03 and then to end up on p.time=0.1 (for writing) p.dt becomes 0.01. We would expect p.dt to return to 0.03 in the following kernel.execute loop. It seems the function [`update_next_dt`](https://github.com/OceanParcels/parcels/blob/master/parcels/particle.py#L230-L236) is written for this purpose. However, p.dt remains 0.01 throughout the rest of the simulation, so we suspect this function does not work as intended. 

In our original simulation the outputdt is a multiple of dt, so this behaviour is not expected, but it might still happen due to rounding errors. We will now further investigate why the p.dt does not get updated and how the p.dt becomes 1E-8 all of the sudden late in a run.
